### ISSUE TITLE
Issue regarding Two way Data Binding (KOTLIN MVVM)

### ISSUE THREAD

**Issue by TharunBalaji2004** [piappstudio/TravelInsurance#14]
Vankkam @piappstudio 🙏, I recently referred your video for two way data binding(https://youtu.be/_-jfB32OYiE) and I actually ended up with an different kind of error while building my app.

Basically, I'm designing a Login screen which consists of two textinputfields `email` and `password`, also I have created a View Model to perform logical form Validation and Authentication. I'm enclosing the source code of ViewModel and XML view.

```kotlin
// LoginViewModel.kt
class LoginViewModel: ViewModel() {

    private val _userEmail = MutableLiveData<String>("")
    val userEmail: LiveData<String> = _userEmail

    private val _userEmailError = MutableLiveData<String?>("")
    val userEmailError: LiveData<String?> = _userEmailError
    
    private val _userPassword = MutableLiveData<String>("")
    val userPassword: LiveData<String> = _userPassword

    private val _userPasswordError = MutableLiveData<String?>("")
    val userPasswordError: LiveData<String?> = _userPasswordError

    fun setUserEmail(userEmail: String){ 
            _userEmail.value = userEmail 
    }
    fun setUserPassword(userPassword: String){ 
           _userPassword.value = userPassword 
    }

    fun loginFormValidation(): Boolean{

        val email = userEmail.value ?: ""
        val password = userPassword.value ?: ""
        var emailValidation = false
        var passwordValidation = false

        if (email.isEmpty()){ _userEmailError.value = "Email address cannot be empty" }
        else if (!isEmailValid(email)){ _userEmailError.value = "Valid Email address is required" }
        else {
            _userEmailError.value = null
            emailValidation = true
        }

        if (password.isEmpty()){ _userPasswordError.value="Password cannot be empty" }
        else {
            _userPasswordError.value = null
            passwordValidation = true
        }

        return emailValidation && passwordValidation
    }
    
    private fun isEmailValid(email: String): Boolean {
        return android.util.Patterns.EMAIL_ADDRESS.matcher(email).matches()
    }
}
```

The source code of Login screen XML view

```kotlin
<layout 
    xmlns:android="http://schemas.android.com/apk/res/android"
    xmlns:app="http://schemas.android.com/apk/res-auto">

    <data>
        <variable
            name="loginViewModel"
            type="com.impetuson.rexroot.viewmodel.LoginViewModel" />
    </data>

    <RelativeLayout
        android:id="@+id/parent_layout"
        android:orientation="vertical"
        android:background="@color/white"
        android:layout_width="match_parent"
        android:layout_height="match_parent">

        <LinearLayout
            android:layout_width="match_parent"
            android:layout_height="match_parent"
            android:layout_below="@+id/ll_header"
            android:gravity="center"
            android:orientation="vertical"
            android:padding="15sp">

            <LinearLayout
                android:layout_width="match_parent"
                android:layout_height="wrap_content"
                android:gravity="center"
                android:orientation="vertical">
                 
                 // Email Field
                <LinearLayout
                    android:layout_width="match_parent"
                    android:layout_height="wrap_content"
                    android:layout_marginBottom="25dp"
                    android:orientation="vertical">

                    <com.google.android.material.textfield.TextInputLayout
                        android:id="@+id/et_email_layout"
                        android:layout_width="match_parent"
                        android:layout_height="wrap_content"
                        app:boxStrokeColor="#2196F3"
                        app:hintTextColor="#2196F3"
                        app:errorEnabled="true"
                        app:error="@{loginViewModel.userEmailError}"
                        app:placeholderText="Eg: john@example.com"
                        app:startIconDrawable="@drawable/baseline_mail_24">

                        <com.google.android.material.textfield.TextInputEditText
                            android:id="@+id/et_email"
                            android:layout_width="match_parent"
                            android:layout_height="wrap_content"
                            android:text="@={loginViewModel.userEmail}"
                            android:hint="Email address" />

                    </com.google.android.material.textfield.TextInputLayout>

                </LinearLayout>
                 
                 // Password Field
                <LinearLayout
                    android:layout_width="match_parent"
                    android:layout_height="wrap_content"
                    android:layout_marginBottom="10dp"
                    android:orientation="vertical">

                    <com.google.android.material.textfield.TextInputLayout
                        android:layout_width="match_parent"
                        android:layout_height="wrap_content"
                        app:boxStrokeColor="#2196F3"
                        app:hintTextColor="#2196F3"
                        app:errorEnabled="true"
                        app:error="@{loginViewModel.userPasswordError}"
                        app:passwordToggleEnabled="true"
                        app:placeholderText="Enter your password"
                        app:startIconDrawable="@drawable/baseline_lock_24">

                        <com.google.android.material.textfield.TextInputEditText
                            android:id="@+id/et_password"
                            android:layout_width="match_parent"
                            android:layout_height="wrap_content"
                            android:text="@={loginViewModel.userPassword}"
                            android:hint="Password"
                            android:inputType="textPassword" />

                    </com.google.android.material.textfield.TextInputLayout>


                </LinearLayout>

                <Button
                    android:id="@+id/btn_login"
                    android:layout_width="wrap_content"
                    android:layout_height="wrap_content"
                    android:paddingTop="14sp"
                    android:paddingBottom="14sp"
                    android:paddingLeft="50dp"
                    android:paddingRight="50dp"
                    android:text="Log in"
                    android:backgroundTint="@color/primary_red"
                    android:textAllCaps="false"
                    android:textColor="@color/white"
                    android:textStyle="bold">
                </Button>

            </LinearLayout>

        </LinearLayout>

    </RelativeLayout>
</layout>
```

When I tried to build my app, I encountered this kind of error saying that the two way binding is not possible here.

```error
The expression 'loginViewModelUserEmail.getValue()' cannot be inverted, so it cannot be used in a two-way binding

Details: There is no inverse for method getValue, you must add an @InverseMethod annotation to the method to indicate which method should be used when using it in two-way binding expressions
```

Kindly look into this issue sir and reply about explaining the problem and solution for this. I rechecked all syntaxes, dependencies and methods included in ViewModel but couldn't figure out what's the real problem here.

This is the `gradle.build` file of the app:
```gradle
plugins {
    alias(libs.plugins.androidApplication)
    alias(libs.plugins.kotlinAndroid)
    id("kotlin-kapt")
}

android {
    namespace = "com.tharunbalaji.loginapp"
    compileSdk = 33

    defaultConfig {
        applicationId = "com.tharunbalaji.loginapp"
        minSdk = 22
        targetSdk = 33
        versionCode = 1
        versionName = "1.0"

        testInstrumentationRunner = "androidx.test.runner.AndroidJUnitRunner"
    }

    buildTypes {
        release {
            isMinifyEnabled = false
            proguardFiles(
                getDefaultProguardFile("proguard-android-optimize.txt"),
                "proguard-rules.pro"
            )
        }
    }
    compileOptions {
        sourceCompatibility = JavaVersion.VERSION_17
        targetCompatibility = JavaVersion.VERSION_17
    }
    kotlinOptions {
        jvmTarget = "17"
    }

    buildFeatures {
        dataBinding = true
    }
}

dependencies {

    implementation(libs.core.ktx)
    implementation(libs.appcompat)
    implementation(libs.material)
    implementation(libs.constraintlayout)
    implementation(libs.androidx.navigation.fragment)
    testImplementation(libs.junit)
    androidTestImplementation(libs.androidx.test.ext.junit)
    androidTestImplementation(libs.espresso.core)

    // Livedata
    implementation("androidx.lifecycle:lifecycle-livedata-ktx:2.3.1")

    // ViewModel
    implementation("androidx.lifecycle:lifecycle-viewmodel-ktx: 2.3.1")
    implementation("androidx.lifecycle:lifecycle-viewmodel-savedstate: 2.3.1")

    // Advanced Navigation - Jetpack Compose
    implementation("androidx.navigation:navigation-fragment-ktx:2.5.3")
    implementation("androidx.navigation:navigation-ui-ktx:2.5.3")
}
```

**Comment by piappstudio** [comment]
Hello Balaji,

Thank you for your email, I think the approach you followed is not the same
as I discussed in my video. In my video, I wasn trying to bind "User"
object into xml, whereas you are trying to bind LiveData<String?> into the
2 way bind.

You can try these 2 things and see what happens,
1. Try to bind MutableLiveData (make is public first) directly instead of
just "LiveData" for a single field property.
2. Use may also try with  ObservableField<String>, if you are dealing with
single field

Or you may follow my same approach like create a separate User class and
have those fields as parameters, expose User as LiveData, and try to bind
it.

Let me know if you have any questions

**Comment by TharunBalaji2004** [comment]
Thank you for your response @piappstudio

As per mentioned in your video, I converted the MutableLiveData and LiveData to refer `LoginModelClass` instead of `String?`, and that solved the error with binding. 

```kotlin
data class LoginModelClass(var userEmail: String,var userPassword: String)
```

Now the issue is solved, thank you for addressing the doubt and your clarification sir! 👍💯
Also, I would like to connect with you through LinkedIn, kindly accept my connection request.

Closing this issue thread
